src/store/reducers/keymap.ts

```diff
@@ -0,0 +1,154 @@
+import '@ngrx/core/add/operator/select';
+import { Action } from '@ngrx/store';
+
+import 'rxjs/add/operator/map';
+import 'rxjs/add/operator/switchMap';
+import { Observable } from 'rxjs/Observable';
+
+import { Keymap } from '../../config-serializer/config-items/Keymap';
+import { KeymapActions } from '../actions';
+import { AppState } from '../index';
+
+const initialState: Keymap[] = [];
+
+export default function(state = initialState, action: Action): Keymap[] {
+    switch (action.type) {
+        case KeymapActions.ADD:
+        case KeymapActions.DUPLICATE:
+            let newKeymap: Keymap = new Keymap;
+            newKeymap = Object.assign(newKeymap, action.payload);
```

Comment:
The keymaps's layers will be shallow copied here. That's why I hate Object.assign. We should use copy constructor everywhere.